File: cellulosesz-api/src/main/java/top/likoslupus/cellulosesz/api/service/ServiceRegistry.java
package top.likoslupus.cellulosesz.api.service;

import java.util.Optional;

public interface ServiceRegistry {

    default <T> Registration register(Class<T> type, T instance) {
        return register(type, instance, "global");
    }

    <T> Registration register(Class<T> type, T instance, String owner);

    <T> T require(Class<T> type);

    <T> Optional<T> optional(Class<T> type);

    boolean contains(Class<?> type);

}

File: cellulosesz-api/src/main/kotlin/top/likoslupus/cellulosesz/api/service/ServiceRegistry.kt
package top.likoslupus.cellulosesz.api.service

public interface ServiceRegistry {

    public fun <T : Any> register(type: Class<T>, instance: T): Registration =
        register(type, instance, "global")

    public fun <T : Any> register(
        type: Class<T>,
        instance: T,
        owner: String
    ): Registration

    public fun <T : Any> require(type: Class<T>): T

    public fun <T : Any> find(type: Class<T>): T?

    public fun contains(type: Class<*>): Boolean

}

public inline fun <reified T : Any> ServiceRegistry.require(): T = require(T::class.java)

public inline fun <reified T : Any> ServiceRegistry.find(): T? = find(T::class.java)

public inline fun <reified T : Any> ServiceRegistry.contains(): Boolean = contains(T::class.java)
